plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

```diff
@@ -0,0 +1,224 @@
+toLspTokenType :: HsSemanticTokenType -> SemanticTokenTypes
+toLspTokenType tk = case tk of
+  -- Function type variable
+  TFunction     -> SemanticTokenTypes_Function
+  -- None function type variable
+  TVariable     -> SemanticTokenTypes_Variable
+  TClass        -> SemanticTokenTypes_Class
+  TClassMethod  -> SemanticTokenTypes_Method
+  TTypeVariable -> SemanticTokenTypes_TypeParameter
+  -- normal data type is a tagged union type look like enum type
+  -- and a record is a product type like struct
+  -- but we don't distinguish them yet
+  TTypeCon      -> SemanticTokenTypes_Enum
+  TDataCon      -> SemanticTokenTypes_EnumMember
+  TRecField     -> SemanticTokenTypes_Property
+  -- pattern syn is like a limited version of macro of constructing a term
+  TPatternSyn   -> SemanticTokenTypes_Macro
+  -- saturated type
+  TTypeSyn      -> SemanticTokenTypes_Type
+  -- not sure if this is correct choice
+  TTypeFamily   -> SemanticTokenTypes_Interface
+
+
+lspTokenReverseMap :: Map.Map SemanticTokenTypes HsSemanticTokenType
+lspTokenReverseMap = Map.fromList $ map (\x -> (toLspTokenType x, x)) $ enumFrom minBound
+
+fromLspTokenType :: SemanticTokenTypes -> Maybe HsSemanticTokenType
+fromLspTokenType tk = Map.lookup tk lspTokenReverseMap
+
+
+-- * 2. Mapping from GHC type and tyThing to semantic token type.
+
+
+-- | tyThingSemantic
+tyThingSemantic :: TyThing -> Maybe HsSemanticTokenType
+tyThingSemantic ty = case ty of
+  AnId vid
+    | isTyVar vid -> Just TTypeVariable
+    | isRecordSelector vid -> Just TRecField
+    | isClassOpId vid -> Just TClassMethod
+    | isFunVar vid -> Just TFunction
+    -- \| isDFunId vid -> TClassMethod
+    | otherwise -> Just TVariable
+  AConLike con -> case con of
+    RealDataCon _ -> Just TDataCon
+    PatSynCon _   -> Just TPatternSyn
+  ATyCon tyCon
+    | isTypeSynonymTyCon tyCon -> Just TTypeSyn
+    | isTypeFamilyTyCon tyCon -> Just TTypeFamily
+    | isClassTyCon tyCon -> Just TClass
+    -- fall back to TTypeCon the result
+    | otherwise -> Just TTypeCon
+  ACoAxiom _ -> Nothing
+  where
+    isFunVar :: Var -> Bool
+    isFunVar var = isFunType $ varType var
+
+isFunType :: Type -> Bool
+isFunType a = case a of
+  ForAllTy _ t    -> isFunType t
+--   Development.IDE.GHC.Compat.Core.FunTy(pattern synonym), FunTyFlag which is used to distinguish
+--   (->, =>, etc..)
+  FunTy flg _ rhs-> isVisibleFunArg flg || isFunType rhs
+  _x              -> isFunTy a
+
+hieKindFunMasksKind :: HieKind a -> HieFunMaskKind a
+hieKindFunMasksKind hieKind = case hieKind of
+  HieFresh -> HieFreshFun
+  HieFromDisk full_file -> HieFromDiskFun $ recoverFunMaskArray (hie_types full_file)
+
+-- wz1000 offered
+-- the idea from https://gitlab.haskell.org/ghc/haddock/-/blob/b0b0e0366457c9aefebcc94df74e5de4d00e17b7/haddock-api/src/Haddock/Backends/Hyperlinker/Utils.hs#L107
+recoverFunMaskArray
+  :: A.Array TypeIndex HieTypeFlat -- ^ flat types
+  -> A.Array TypeIndex Bool-- ^ full AST
```

Comment:
comments generally seem inaccurate for what this does? (looking for which types are functions)

Reply:
thanx, fixed